pkg/apis/core/validation/validation.go

```diff
@@ -5454,28 +5460,28 @@ func validateScopedResourceSelectorRequirement(resourceQuotaSpec *core.ResourceQ
 	}
 	for _, invalidScopePair := range invalidScopePairs {
 		if scopeSet.HasAll(invalidScopePair.List()...) {
-			allErrs = append(allErrs, field.Invalid(fldPath, resourceQuotaSpec.Scopes, "conflicting scopes"))
+			allErrs = append(allErrs, field.Invalid(fldPath, newSpec.Scopes, "conflicting scopes"))
 		}
 	}
 
 	return allErrs
 }
 
 // validateScopeSelector tests that the specified scope selector has valid data
-func validateScopeSelector(resourceQuotaSpec *core.ResourceQuotaSpec, fld *field.Path) field.ErrorList {
+func validateScopeSelector(newSpec, oldSpec *core.ResourceQuotaSpec, fld *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if resourceQuotaSpec.ScopeSelector == nil {
+	if newSpec.ScopeSelector == nil {
 		return allErrs
 	}
-	allErrs = append(allErrs, validateScopedResourceSelectorRequirement(resourceQuotaSpec, fld.Child("scopeSelector"))...)
+	allErrs = append(allErrs, validateScopedResourceSelectorRequirement(newSpec, oldSpec, fld.Child("scopeSelector"))...)
 	return allErrs
 }
 
 // ValidateResourceQuota tests if required fields in the ResourceQuota are set.
 func ValidateResourceQuota(resourceQuota *core.ResourceQuota) field.ErrorList {
 	allErrs := ValidateObjectMeta(&resourceQuota.ObjectMeta, true, ValidateResourceQuotaName, field.NewPath("metadata"))
 
-	allErrs = append(allErrs, ValidateResourceQuotaSpec(&resourceQuota.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateResourceQuotaSpec(&resourceQuota.Spec, nil, field.NewPath("spec"))...)
```

Comment:
`ValidateResourceQuota` is called by `resourcequotaStrategy#ValidateUpdate`, which means an API update of an object using the new scope with the feature off will have stricter validation applied to it incorrectly.

I'd suggest using the same approach used for pods:
* a ResourceQuotaValidationOptions struct with an `AllowPodAffinityNamespaceSelector` field
* a `GetValidationOptionsFromResourceQuota(newObj, oldObj ResourceQuota) ResourceQuotaValidationOptions` helper that sets that field based on the feature enablement and whether the old object is non-nil and used the scope
* Change `ValidateResourceQuota` and `ValidateResourceQuotaUpdate` to take the options struct and plumb it to the validateResourceQuotaScopes / validateScopedResourceSelectorRequirement / validateScopeSelector functions (this step would have caught the unexpected call to ValidateResourceQuota from `resourcequotaStrategy#ValidateUpdate`
* update `resourcequotaStrategy#Validate` and `resourcequotaStrategy#ValidateUpdate` to pass the result of calling the helper method to the validation functions

Reply:
sg discussed offline, the logic is already implementing the correct semantics, but I refactored it to use the opts pattern.